### ISSUE TITLE
memory leaks in t220

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#1428]
```
$ make ASAN=1
$ cd tests
$ ./runtest.py -vdp -O0 220
Start 1 tests with 1 worker
Test case                 pg
------------------------: O0
build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -pg -O0  s-abc.c
test command: /home/honggyu/work/uftrace/git/uftrace/uftrace live --no-pager --no-event --libmcount-path=/home/honggyu/work/uftrace/git/uftrace --force -F fork -F main ./test-script.sh

=================================================================
==20593==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1024 byte(s) in 1 object(s) allocated from:
    #0 0x7f977bd85ca8 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x10bca8)
    #1 0x568d3c in check_script_file /home/honggyu/work/uftrace/git/uftrace/utils/symbol.c:186
    #2 0x498d25 in check_binary /home/honggyu/work/uftrace/git/uftrace/cmds/record.c:1610
    #3 0x4a1bf6 in command_record /home/honggyu/work/uftrace/git/uftrace/cmds/record.c:2204
    #4 0x4699d6 in command_live /home/honggyu/work/uftrace/git/uftrace/cmds/live.c:149
    #5 0x411e44 in main /home/honggyu/work/uftrace/git/uftrace/uftrace.c:1443
    #6 0x7f9779dcb0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: 1024 byte(s) leaked in 1 allocation(s).
t220_trace_script: -pg -O0 returns 1

220 trace_script        : NZ
```

**Comment by japawBlob** [comment]
Hey I think I found the missing free. Test 220 no longer leaks in my fork. Will do some further tests.

**Comment by honggyukim** [comment]
@japawBlob Thanks! Please send a PR when you're ready.

**Comment by honggyukim** [comment]
@japawBlob Sorry but @AndrewWoo7 sent a PR before your submission so I will accept his patch. Thanks anyway.

**Comment by japawBlob** [comment]
No worries! I am sorry for ghosting you out. My computer broke and I still didn't got to fix it to submit my code. Thanks for update and take care.

**Comment by honggyukim** [comment]
Thanks. I hope to have another chance to get your patch next time. Take care you too!

**Comment by honggyukim** [comment]
I should reopen this issue because #1479 is reverted at #1483.

**Comment by JHH20** [comment]
I can think of some solutions to freeing `uftrace_opts.exename` but maybe someone might have a better idea?

| Idea | Notes|
| --- | --- |
| Add a boolean / bit flag to `struct uftrace_opts` for storing whether a ptr was malloc-ed or not | Clutters the struct as more ptrs need this check. If bit flag is used, an enum for bitmasks should also be added |
| Modify `uftrace_opts.exename` to be a `struct {char *ptr, boolean needsFree}` | Adds another layer of indirection, and existing references must be updated |
| Add a ptr array to `struct uftrace_opts` for storing ptrs that need to be freed | If array size is set to number of ptr fields, struct size increases. Otherwise, deal with dynamic re-allocation whenever adding a ptr to the array |
| Search near the ptr for metadata to tell if the ptr was malloc-ed | Not portable since malloc/free are implementation dependent. Search attempt may cause segfault |
| Use `__builtin_frame_address` macro at `main()` and `free_opts()` to check if ptr is on the stack | Fails if allocation occurs on the stack (eg `alloca()`) and relies on non-standard macro which may break on certain environments. Also, may need different solution to find the stack base if some stack var exist outside `main()` stack frame (eg: argv) |
| Define wrapper for malloc and free such that malloc will track allocated address and free will safely ignore non-allocated addresses | Bad effort-to-return ratio to do it right |
| Always malloc (memcpy) | Waste memory |

From the list above, I think 3 is probably the most feasible. Interacting with the ptr array can be abstracted in a function. It would not be necessary to store the array length in the `struct uftrace_opts` since that can be done within the function (either null-terminate the array, or use static local var to track last index)

Downside is that 2 and 3 change the allocation primitive so it may be easy to forget doing the right thing further down the line. Also since C isn't that object-oriented, using a setter function to enforce this policy would be unnatural.

**Comment by honggyukim** [comment]
@JHH20 @japawBlob Thanks a lot for your consideration! But I'd rather make a simpler solution. How about this?
```diff
diff --git a/cmds/record.c b/cmds/record.c
index 54f8ce9b..4477b933 100644
--- a/cmds/record.c
+++ b/cmds/record.c
@@ -1605,6 +1605,7 @@ again:
                if (p)
                        *p = '\0';

+               free(opts->exename);
                opts->exename = script;
                close(fd);
                goto again;
diff --git a/uftrace.c b/uftrace.c
index 4d787286..a1a5b469 100644
--- a/uftrace.c
+++ b/uftrace.c
@@ -1106,13 +1106,13 @@ static void parse_opt_file(int *argc, char ***argv, char *filename, struct uftra
                *argv = file_argv;

                opts->idx = optind;
-               opts->exename = file_argv[optind];
+               opts->exename = xstrdup(file_argv[optind]);

                /* mark it to free at the end */
                opts->opt_file = filename;
        }
        else {
-               opts->exename = orig_exename;
+               opts->exename = xstrdup(orig_exename);
                free_parsed_cmdline(file_argv);
        }

@@ -1219,6 +1219,7 @@ static void free_opts(struct uftrace_opts *opts)
        free(opts->caller);
        free(opts->watch);
        free(opts->hide);
+       free(opts->exename);
        free_parsed_cmdline(opts->run_cmd);
 }

@@ -1250,7 +1251,7 @@ static int parse_options(int argc, char **argv, struct uftrace_opts *opts)

        if (optind < argc) {
                opts->idx = optind;
-               opts->exename = argv[optind];
+               opts->exename = xstrdup(argv[optind]);
        }

        return 0;
```
The approach is simple. We used to have a mixed memory for `opts->exename`. This change makes `opts->exename` always contains freeable memory.

**Comment by honggyukim** [comment]
This was fixed by 9c8d108af51e5dd1994a0a3b685773639742798c and the leak is no longer exist. I will close it.
